### DISCUSSION TITLE
Converting From using Image to Mat

### DISCUSSION THREAD

**Discussion by PPS100** [emgucv/emgucv#535] [Q&A]
I am upgrading a project to the latest version of Emgu. Wanted some guidance on the correct way to convert from Image to Mat. I want to make sure that I do not break my functionality in doing so. 

Below is a example of code that I would like to move from using Image to Mat. Wanted to see if some could show me how they would update this code. I am worried that since I am loosing the descriptors "Tcc" and "Gray" that things will break. 

protected Image<Gray, Byte> CreateThresholdImageYcc<TColor, TDepth>(Image<TColor, TDepth> image, int threshold, int smoothing)
            where TColor : struct, Emgu.CV.IColor
            where TDepth : new()
        {
            Image<Gray, Byte> thresholdImage = null;

            // Convert image to YCC color space and split into separate channels
            // Equalize histograms of channels 1 and 2, and combine to emphasize football
            using (Image<Ycc, Byte> yccImage = Emgu.CV.BitmapExtension.ToImage<Ycc, Byte>(TemplateImage))
            {
                Image<Gray, Byte>[] yccChannels = yccImage.Split();
                Image<Gray, Byte> imageToThreshold = null;

                try
                {
                    yccChannels[1]._EqualizeHist();
                    yccChannels[2]._EqualizeHist();

                    using (Image<Gray, Byte> ycc2Not = yccChannels[2].Not())
                    using (Image<Gray, Byte> ycc1AndYcc2Not = yccChannels[1].And(ycc2Not))
                    {
                        // Set image to threshold
                        
                        if (smoothing > 0)
                        {
                            imageToThreshold = ycc1AndYcc2Not.SmoothGaussian((smoothing * 2) - 1);
                        }
                        else
                        {
                            imageToThreshold = ycc1AndYcc2Not;
                        }
                        
                        thresholdImage = imageToThreshold.ThresholdBinary(new Gray(threshold), new Gray(255));
                    }
                }
                finally
                {
                    for (int i = 0; i < yccChannels.Count(); i++)
                    {
                        yccChannels[i].Dispose();
                    }

                    if (imageToThreshold != null)
                    {
                        imageToThreshold.Dispose();
                    }
                }
            }

            return thresholdImage;

**Comment by sn4k3** [comment]
Does that work to you?

```C#
public Mat CreateThresholdImageYcc(Mat src, int threshold, int smoothing)
{
    Mat imageToThreshold = null;

    using (var mat = new Mat())
    {
        CvInvoke.CvtColor(src, mat, ColorConversion.Bgr2YCrCb);

        var matChannels = mat.Split();
        if (matChannels.Length < 3) return null;

        try
        {
            CvInvoke.EqualizeHist(matChannels[1], matChannels[1]);
            CvInvoke.EqualizeHist(matChannels[2], matChannels[2]);

            imageToThreshold = new Mat();

            using (var ycc2Not = new Mat())
            using (var ycc1AndYcc2Not = new Mat())
            {
                CvInvoke.BitwiseNot(matChannels[2], ycc2Not);
                CvInvoke.BitwiseAnd(matChannels[1], ycc2Not, ycc1AndYcc2Not);
                // Set image to threshold

                if (smoothing > 2)
                {
                    smoothing = (smoothing * 2) - 1;
                    CvInvoke.GaussianBlur(ycc1AndYcc2Not, imageToThreshold, new Size(smoothing, smoothing), 0);
                }
                else
                {
                    imageToThreshold = ycc1AndYcc2Not;
                }

                CvInvoke.Threshold(imageToThreshold, imageToThreshold, threshold, byte.MaxValue, ThresholdType.Binary);
            }

        }
        finally
        {
            for (int i = 0; i < matChannels.Length; i++)
            {
                matChannels[i].Dispose();
            }
        }
    }

    return imageToThreshold;
}
```
